Documentation For Simulacra (#1337)
GoogleCloudPlatform/ops-agent: cmd/simulacra/README.md

# Simulacra 

> Simulacra is a command line tool that the oncall team can use to automatically spin up VMs with a customers environment. 

Reviewer: ```suggestion
Simulacra is a command line tool that Ops Agent team members can use to spin up VMs with specific Ops Agent versions, third party applications, and any custom additional scripts. 
```

> Simulacra can read Ops Agent config files and infer all the third party apps that it needs to install. It then instantiates the VM according to user specifications, installs Ops Agent and then installs all the required third party applications. 

Reviewer: This paragraph being one of the first things said is a bit confusing. I'd start with something more basic like "Simulacra can be configured through command line for quick operations, or a config file for a reproducible setup" and then go into how it can also read a config file to infer apps.

> We can specify various configurations for our desired VM. For example, we can specify the Operating System we want our VM to have. 

Reviewer: I think this can be deleted as it will be clear from the future sections.

> Ultimately, we want to integrate Sigmulacra with our diagnostic tool and be able to generate Simulacra config files from the output of the diagnostic tool.  

Reviewer: I think this can be deleted since it doesn't need to go in the README.

## Example Usage

    > cat config.yaml
        logging:
            receivers:
                redis:
                    type: redis
            service:
                pipelines:
                    redis:
                        receivers:
                            - redis

    go run -tags=integration_test simulacra.go --ops_agent_config "/usr/local/google/home/shafinsiddique/ops-agent/integration_test/simulacra/config.yaml"

> This command invokes Simulacra to spin up a VM, install Ops Agent on it and then install the required third party applications. Based on the ops agent config file, the only third party application Ops Agent is configured with is Redis so Simulacra will only install Redis. 

Reviewer: ```suggestion
This command invokes Simulacra to spin up a VM, install the Ops Agent on it and then install the required third party applications. Based on the Ops Agent config in the example above, Simulacra will install Redis on the created VM. 
```

## Command Line Arguments

> For any advanced usage, Simulacra should be configured using a YAML file and then invoked by specifying the --config argument. 

Reviewer: I would delete this line. Specifying that this is for quick one-offs in the next line is sufficient.

> However, we also offer a subset of the total config parameters as command line arguments. This is for quick usages. 

Reviewer: ```suggestion
We offer a subset of the possible configuration parameters as command line arguments. This is for quick one-off usage. 
```

    --platform: Optional. The OS for the VM. If missing, debian-11 is used.

    --ops_agent_config: Optional. Path to the Ops Agent Config File. If unspecified, Ops Agent will not install any third party applications and configure Ops Agent with default settings.

    --project: Optional. The project ID for the project that will be used to instantiate the VMs. If missing, Simulacra will try to infer from GCloud config.

    --zone: Optional. The zone in which the VM will be instantiated in. If missing, Simulacra will try to infer from GCloud config.

    --name: Optional. A name for the instance to be created. If missing, a random name with prefix 'simulacra-vm-instance' will be assigned.

    --third_party_apps_path: Optional. The path to the third party apps data folder. If missing, Simulacra assumes the working directory is the root of the repo. Therefore, the default path is './integration_test/third_party_apps_data'

## YAML Config

> As mentioned above, Simulacra can be configured by creating a YAML file and then specifying the '--config' command line argument. 

Reviewer: ```suggestion
Simulacra can be configured by creating a YAML file and then specifying the '--config' command line argument. 
```

> Using YAML files, we can get even more options for configuring our VMs than just using command line arguments. 

Reviewer: I'd delete this line, mentioning above that the command line is a subset is sufficient to know that the YAML config format is where the full config surface is. If you'd like, you could just simply specify that the YAML config contains the full breadth of options for configuring Simulacra.

>     - platform (String): Optional. The OS for the VM. If missing, debian-11 is used.

Reviewer: For better or worse we use the word `platform` generally in place of `OS` in most spots in this repo so we can just stay consistent with that
```suggestion
    - platform (String): Optional. The platform for the VM. If missing, debian-11 is used.
```

>     -- ops_agent_config (String): Optional. The OS for the VM. If missing, debian-11 is used.

Reviewer: ```suggestion
    - ops_agent_config (String): Optional. The OS for the VM. If missing, `debian-11` is used.
```

>     - project (String): Optional. The project ID for the project that will be used to instantiate the VMs. If missing, Simulacra will try to infer from GCloud config.

Reviewer: ```suggestion
    - project (String): Optional. The project ID for the project where the VMs will be created. If missing, Simulacra will try to infer from GCloud config.
```

>     --zone (String): Optional. The zone in which the VM will be instantiated in. If missing, Simulacra will try to infer from GCloud config.

Reviewer: ```suggestion
    - zone (String): Optional. The zone where the VM will be created. If missing, Simulacra will try to infer from GCloud config.
```
